### PR TITLE
Replace COPS=2 with CFUN=0 (or CFUN=4) to prevent longer registration times

### PR DESCRIPTION
**Pull request by keeramis**
# Problem

Cell registration takes a long time with certain SIMs in certain locations after using `Cellular.disconnect()` or reset through `particle usb reset`.

### Solution
Gen2 and Gen3: 
- `disconnect()` currently uses `COPS=2` which presumably initiates some kind of rescan on the modem for network, and takes longer time when it's trying to register. Setting the `QCFG` parameters on Quectel modem could also re-trigger scan that could lead to longer reg times.

Gen3:
- `COPS=2` is also done in `initReady()` which again delays the registration. This can be removed as per some testing, `COPS=2` did not look like it is needed. (Please see [this](https://github.com/particle-iot/device-os/pull/2175/files#r473053805) comment) 

Gen2: 
- `COPS=2` is used for setting `CGDCONT` and `URAT` which requires de-registration before being able to set them. This can also be achieved by `AT+CFUN=0` or `AT+CFUN=4` which can be used to de-register before setting these commands, instead of using `COPS=2` with which registration time lengthens. 

### Steps to Test

Check that the connection at Step3 will not take longer than the connection at Step1. 
Do not use `particle usb reset` to reset the module to stat this test, as it triggers as `disconnect()`. 

### Example App

```c
#include "application.h"

SerialLogHandler g_logUsbSerial(LOG_LEVEL_ALL);
SYSTEM_MODE(AUTOMATIC);

void setup(){
    // Step 1: Let a first connection work,
    // and make sure the device is powered on for the first time
    Particle.connect();
    waitUntil(Particle.connected);
    Serial.printlnf("Particle connected");

    // Step 2: Disconnect from cloud and cellular network
    Particle.disconnect();
    waitUntil(Particle.disconnected);
    Cellular.disconnect();
    while(Cellular.ready());

    Serial.printlnf("Cellular disconnected");
    delay(20s);

    // Step 3: 
    Cellular.connect();
    waitUntil(Cellular.ready);
    Serial.printlnf("Cellular connected");
}

void loop() {
}
```

### References

[ch59733](https://app.clubhouse.io/particle/story/59733/quantify-registration-times-in-brn402-between-twilio-and-at-t-sims)

---

### Completeness

- [x] User is totes amazing for contributing!
- [x] Contributor has signed CLA ([Info here](https://github.com/spark/firmware/blob/develop/CONTRIBUTING.md))
- [x] Problem and Solution clearly stated
- [ ] Run unit/integration/application tests on device
- [ ] Added documentation
- [ ] Added to CHANGELOG.md after merging (add links to docs and issues)

### CODE REVIEW

File: hal/network/ncp_client/quectel/quectel_ncp_client.cpp [hunk 1]
```diff
@@ -964,10 +964,6 @@ int QuectelNcpClient::initReady(ModemState state) {
         // Select either internal or external SIM card slot depending on the configuration
         CHECK(selectSimCard());
 
-        // Just in case disconnect
-        int r = CHECK_PARSER(parser_.execCommand("AT+COPS=2"));
-        // CHECK_TRUE(r == AtResponse::OK, SYSTEM_ERROR_UNKNOWN);
-
```
Comment (keeramis) [inline hal/network/ncp_client/quectel/quectel_ncp_client.cpp:970]: Example for longer reg time with and without using `COPS=2` on `EG91`:

- With `COPS=2`

```
0000016015 [ncp.at] TRACE: > AT+CGDCONT=1,"IP","super"
0000016047 [ncp.at] TRACE: < OK
0000016048 [ncp.at] TRACE: > AT+CFUN=1,0
0000016051 [ncp.at] TRACE: < OK
0000016052 [ncp.at] TRACE: > AT+CREG=2
0000016054 [ncp.at] TRACE: < OK
0000016055 [ncp.at] TRACE: > AT+CGREG=2
0000016059 [ncp.at] TRACE: < OK
0000016060 [ncp.at] TRACE: > AT+CEREG=2
0000016065 [ncp.at] TRACE: < OK
0000016066 [ncp.client] TRACE: NCP connection state changed: 1
0000016076 [net.pppncp] TRACE: NCP event 2
0000016077 [net.pppncp] TRACE: State changed event: 1
0000016082 [ncp.at] TRACE: > AT+COPS?
0000016088 [ncp.at] TRACE: < +COPS: 2
0000016088 [ncp.at] TRACE: < OK
0000016093 [ncp.at] TRACE: > AT+COPS=0,2
0000016082 [net.ppp.client] TRACE: PPP thread event LOWER_DOWN
0000016111 [ncp.at] TRACE: < OK
0000016112 [ncp.at] TRACE: > AT+CREG?
0000016115 [ncp.at] TRACE: < +QIND: SMS DONE
0000016116 [ncp.at] TRACE: < +CREG: 2,0
0000016117 [ncp.at] TRACE: < OK
0000016124 [ncp.at] TRACE: > AT+CGREG?
0000016127 [ncp.at] TRACE: < +CGREG: 2,0
0000016128 [ncp.at] TRACE: < OK
0000016129 [ncp.at] TRACE: > AT+CEREG?
0000016135 [ncp.at] TRACE: < +CEREG: 2,0
0000016136 [ncp.at] TRACE: < OK
0000031140 [ncp.at] TRACE: > AT+CEER
0000031145 [ncp.at] TRACE: < +CEER: 0,-1
0000031146 [ncp.at] TRACE: < OK
0000031147 [ncp.at] TRACE: > AT+CREG?
0000031149 [ncp.at] TRACE: < +CREG: 2,2
0000031152 [ncp.at] TRACE: < OK
0000031153 [ncp.at] TRACE: > AT+CGREG?
0000031156 [ncp.at] TRACE: < +CGREG: 2,2
0000031164 [ncp.at] TRACE: < OK
0000031165 [ncp.at] TRACE: > AT+CEREG?
0000031168 [ncp.at] TRACE: < +CEREG: 2,2
0000031175 [ncp.at] TRACE: < OK
0000046175 [ncp.at] TRACE: > AT+CEER
0000046178 [ncp.at] TRACE: < +CEER: 0,-1
0000046179 [ncp.at] TRACE: < OK
0000046180 [ncp.at] TRACE: > AT+CREG?
0000046182 [ncp.at] TRACE: < +CREG: 2,2
0000046183 [ncp.at] TRACE: < OK
0000046189 [ncp.at] TRACE: > AT+CGREG?
0000046192 [ncp.at] TRACE: < +CGREG: 2,2
0000046194 [ncp.at] TRACE: < OK
0000046200 [ncp.at] TRACE: > AT+CEREG?
0000046203 [ncp.at] TRACE: < +CEREG: 2,2
0000046205 [ncp.at] TRACE: < OK
0000054205 [ncp.at] TRACE: < +CREG: 5,"6A41","7F3CC11",7
0000054306 [ncp.at] TRACE: < +CGREG: 5,"6A41","7F3CC11",7
```

- Without `COPS=2` (shorter!)

```
0000067363 [ncp.at] TRACE: > AT+CGDCONT=1,"IP","super"
0000067393 [ncp.at] TRACE: < OK
0000067394 [ncp.at] TRACE: > AT+CFUN=1,0
0000067396 [ncp.at] TRACE: < OK
0000067397 [ncp.at] TRACE: > AT+CREG=2
0000067399 [ncp.at] TRACE: < OK
0000067400 [ncp.at] TRACE: > AT+CGREG=2
0000067407 [ncp.at] TRACE: < OK
0000067408 [ncp.at] TRACE: > AT+CEREG=2
0000067411 [ncp.at] TRACE: < OK
0000067411 [ncp.client] TRACE: NCP connection state changed: 1
0000067419 [net.pppncp] TRACE: NCP event 2
0000067422 [net.pppncp] TRACE: State changed event: 1
0000067430 [net.ppp.client] TRACE: PPP thread event LOWER_DOWN
0000067431 [ncp.at] TRACE: > AT+COPS?
0000067441 [ncp.at] TRACE: < +COPS: 0,0,"AT&T Twilio",7
0000067442 [ncp.at] TRACE: < OK
0000067445 [ncp.at] TRACE: > AT+CREG?
0000067452 [ncp.at] TRACE: < +CREG: 2,5,"6A41","7F46703",7
0000067453 [ncp.at] TRACE: < OK
0000067456 [ncp.at] TRACE: > AT+CGREG?
0000067463 [ncp.at] TRACE: < +CGREG: 2,5,"6A41","7F46703",7
0000067465 [ncp.at] TRACE: < OK
0000067467 [ncp.at] TRACE: > AT+CEREG?
0000067474 [ncp.at] TRACE: < +CEREG: 2,5,"6A41","7F46703",7
0000067475 [ncp.at] TRACE: < OK
```

**Comment by keeramis** [comment]
[Just for reference]
An example log on `EG91` on older disconnect() vs this PR's disconnect() function: 
[EG91_COPS=2_vs_CFUN=0.txt](https://github.com/particle-iot/device-os/files/5105420/EG91_COPS.2_vs_CFUN.0.txt)

File: hal/src/electron/modem/mdm_hal.cpp [hunk 2]
```diff
@@ -1275,6 +1286,18 @@ int MDMParser::_cbURAT(int type, const char *buf, int len, bool *matched_default
 bool MDMParser::registerNet(const char* apn, NetStatus* status, system_tick_t timeout_ms)
 {
     LOCK();
+
+    // Set to full functionality mode
+    int cfun_val;
```
Comment (avtolstoy) [inline hal/src/electron/modem/mdm_hal.cpp:1291]: Should be initialized to some default value, as `_cbCFUN` _might_ not modify it and leave it at an unknown value.

File: hal/src/electron/modem/mdm_hal.cpp [hunk 3]
```diff
@@ -1362,6 +1385,17 @@ bool MDMParser::registerNet(const char* apn, NetStatus* status, system_tick_t ti
             if (!_atOk()) {
                 goto failure;
             }
+
+            int cfun_val;
```
Comment (avtolstoy) [inline hal/src/electron/modem/mdm_hal.cpp:1389]: Initialize to default value.

**Review by avtolstoy — CHANGES_REQUESTED** [review]

File: hal/src/electron/modem/mdm_hal.cpp (hunk 2 above)
Comment (keeramis) [reply hal/src/electron/modem/mdm_hal.cpp:1291]: Very true. thank u

File: hal/src/electron/modem/mdm_hal.cpp [hunk 4]
```diff
@@ -1275,6 +1286,18 @@ int MDMParser::_cbURAT(int type, const char *buf, int len, bool *matched_default
 bool MDMParser::registerNet(const char* apn, NetStatus* status, system_tick_t timeout_ms)
 {
     LOCK();
+
+    // Set to full functionality mode
+    int cfun_val = -1;
+    sendFormated("AT+CFUN?\r\n");
+    if (RESP_OK != waitFinalResp(_cbCFUN, &cfun_val))
+        goto failure;
```
Comment (avtolstoy) [inline hal/src/electron/modem/mdm_hal.cpp:1294]: Nit: prefer to always use `{ }` even for single-line statements.

**Review by avtolstoy — APPROVED** [review]

**Pull request merged**